serde_derive/src/ser.rs

```diff
@@ -877,20 +924,33 @@ fn serialize_struct_visitor(
             }
 
             let span = Span::def_site().located_at(field.original.span());
-            let func = struct_trait.serialize_field(span);
-            let ser = quote! {
-                try!(#func(&mut __serde_state, #key_expr, #field_expr));
+            let ser = if field.attrs.flatten() {
```

Comment:
This fails to compile if we are inside a struct variant. Could you handle that case if it is easy, or produce an error instead?

```rust
#[derive(Serialize)]
enum E {
    Q {
        #[serde(flatten)]
        flat: BTreeMap<String, String>,
    },
}
```

```
error[E0277]: the trait bound `<__S as serde::Serializer>::SerializeStructVariant: serde::ser::SerializeMap` is not satisfied
 --> src/main.rs:9:10
  |
9 | #[derive(Serialize)]
  |          ^^^^^^^^^ the trait `serde::ser::SerializeMap` is not implemented for `<__S as serde::Serializer>::SerializeStructVariant`
  |
  = help: consider adding a `where <__S as serde::Serializer>::SerializeStructVariant: serde::ser::SerializeMap` bound
  = note: required because of the requirements on the impl of `serde::Serializer` for `serde::private::ser::FlatMapSerializer<'_, <__S as serde::Serializer>::SerializeStructVariant>`
```